lib/core/src/Control/Concurrent/Concierge.hs

```diff
@@ -0,0 +1,85 @@
+{-# LANGUAGE NamedFieldPuns #-}
+{-# LANGUAGE RankNTypes #-}
+-- |
+-- Copyright: © 2021 IOHK
+-- License: Apache-2.0
+--
+-- This module provides a utility for ordering concurrent actions
+-- via locks.
+module Control.Concurrent.Concierge
+    ( Concierge
+    , newConcierge
+    , atomicallyWith
+    , atomicallyWithLifted
+    )
+    where
+
+import Prelude
+
+import Control.Monad.Class.MonadFork
+    ( MonadThread, ThreadId, myThreadId )
+import Control.Monad.Class.MonadSTM
+    ( MonadSTM
+    , TVar
+    , atomically
+    , modifyTVar
+    , newTVarIO
+    , readTVar
+    , retry
+    , writeTVar
+    )
+import Control.Monad.Class.MonadThrow
+    ( MonadThrow, bracket )
+import Control.Monad.IO.Class
+    ( MonadIO, liftIO )
+import Data.Map.Strict
+    ( Map )
+
+import qualified Data.Map.Strict as Map
+
+{-------------------------------------------------------------------------------
+    Concierge
+-------------------------------------------------------------------------------}
+-- | At a 'Concierge', you can obtain a lock and
+-- enforce sequential execution of concurrent 'IO' actions.
+--
+-- Back in the old days, hotel concierges used to give out keys.
+-- But after the cryptocurrency revolution, they give out locks. :)
+-- (The term /lock/ is standard terminology in concurrent programming.)
```

Comment:
Sounds like you're drawing a long bow with this Concierge name. :stuck_out_tongue:

Reply:
:joy: I have a sweet spot for figurative names. I briefly considered "Locksmith", but that suggested that the locks were being created on the fly instead of being rented out. The drawback of figurative names is that they take a moment of explanation, but the drawback of generic names like "lock collection", "lock factory", "lock registry", … is that it's easy to forget what they do.